### PR TITLE
Feature/sign up form

### PR DESCRIPTION
**Pull request by Bluesmile82**
This PR adds the signup page. It branches from feature/registration_endpoint
WIP - Until that branch is merged
The server returns a 422 unprocessable entity when the email is already in the database.

- Create signup link in login page and route
- Create signup page
- Handle frontend validation of the form
- Print message in login screen if succeded
image

### CODE REVIEW

File: app/javascript/app/pages/sign-up/sign-up-component.jsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          <Input
+            key={i.slug}
+            label={i.label}
+            inputType={i.type}
+            placeholder={i.placeholder}
+            onChange={value => handleValueChange(i.slug, value)}
+            onBlur={value => handleValueChange(i.slug, value)}
+          />
+        ))}
+        <Button onClick={handleSubmit} theme={yellowButtonTheme}>
```
Comment (aabdaab) [inline app/javascript/app/pages/sign-up/sign-up-component.jsx:34]: don't we have already the props in Button accepting color? like `<Button color={'yellow'} onClick...>`

File: app/javascript/app/utils/user-validation.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+export default user => {
+  const allFieldsFilled = Object.keys(user).every(k => user[k]);
+  const passConfirmationIsOk = user.password === user.password_confirmation;
+  const emailregex = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+  const emailValid = user.email && user.email.match(emailregex);
```
Comment (aabdaab) [inline app/javascript/app/utils/user-validation.js]: I'd move it to another file `email-validation` and `password-validation` that would implement `validate` function and use it like so:
```
import emailValidator from 'email-validation';

const isEmailValid = emailValidator.validate(user.email)
```
also wdyt about changing the var names to more readable, like isEmailValid, isPasswordValid, etc?

**Review by aabdaab — CHANGES_REQUESTED** [review]

File: app/javascript/app/pages/sign-up/sign-up-component.jsx (hunk 1 above)
Comment (Bluesmile82) [reply app/javascript/app/pages/sign-up/sign-up-component.jsx:34]: That is in CW. I prefer the theme approach. Also sometimes there are more changes apart from the color

Comment (aabdaab) [reply app/javascript/app/pages/sign-up/sign-up-component.jsx:34]: aaah, you're right. different repo, hah. in that case, I agree, I like the theme approach better 👍

**Comment by agnessa** [comment]
@Bluesmile82 I had a look again at the duplicate email issue, indeed the endpoint returns 422 but also a payload: `{"errors":{"email":["has already been taken"]}}` - can you access it, does it solve the problem?

**Comment by Bluesmile82** [comment]
Sure, that will help. Thanks, Agnieszka!

**Comment by Bluesmile82** [comment]
Ready again!

**Review by aabdaab — APPROVED** [review]

File: app/javascript/app/services/login.service.js [hunk 3]
```diff
@@ -30,9 +30,8 @@ export async function signUp(user) {
       'Content-Type': 'application/json',
       Accept: 'application/json'
     })
-  }).then(function (response) {
-    if (response.ok) return true;
-    if (response.status === 422) return false;
+  }).then(response => {
+    if (response.ok || response.status === 422) return response.json();
```
Comment (weberjavi) [inline app/javascript/app/services/login.service.js]: Just missing some comment to future developers explaining why `422` should be a valid response. 
Does it makes sense?

Comment (agnessa) [reply app/javascript/app/services/login.service.js]: Hi! It makes sense according to its definition: https://httpstatuses.com/422; there are alternatives possible (e.g. 400), this is the convention adopted by Rails.

**Pull request merged**